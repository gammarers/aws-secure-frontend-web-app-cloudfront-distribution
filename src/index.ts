import * as cdk from 'aws-cdk-lib';
import * as acm from 'aws-cdk-lib/aws-certificatemanager';
import * as cloudfront from 'aws-cdk-lib/aws-cloudfront';
import * as origins from 'aws-cdk-lib/aws-cloudfront-origins';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

export interface SecureFrontendWebAppCloudFrontDistributionProps {
  readonly comment?: string;
  readonly domainName: string;
  readonly certificate: acm.ICertificate;
  readonly originBucket: s3.IBucket;
  readonly accessLogBucket?: s3.IBucket;
  readonly priceClass?: cloudfront.PriceClass;
}

export class SecureFrontendWebAppCloudFrontDistribution extends cloudfront.Distribution {

  // eslint-disable-next-line max-len
  constructor(scope: Construct, id: string, props: SecureFrontendWebAppCloudFrontDistributionProps) {
    super(scope, id, {
      enabled: true,
      comment: props.comment,
      defaultRootObject: 'index.html',
      certificate: props.certificate,
      domainNames: [props.domainName],
      minimumProtocolVersion: cloudfront.SecurityPolicyProtocol.TLS_V1_2_2021,
      sslSupportMethod: cloudfront.SSLMethod.SNI,
      httpVersion: cloudfront.HttpVersion.HTTP2_AND_3,
      // webAclId: props.wafAclId,
      logBucket: props.accessLogBucket,
      logFilePrefix: props.accessLogBucket ? `${props.domainName}/` : undefined,
      defaultBehavior: {
        allowedMethods: cloudfront.AllowedMethods.ALLOW_GET_HEAD_OPTIONS,
        cachedMethods: cloudfront.CachedMethods.CACHE_GET_HEAD_OPTIONS,
        cachePolicy: cloudfront.CachePolicy.CACHING_OPTIMIZED,
        compress: true,
        viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.HTTPS_ONLY,
        origin: origins.S3BucketOrigin.withOriginAccessControl(props.originBucket),
        responseHeadersPolicy: new cloudfront.ResponseHeadersPolicy(scope, 'ResponseHeadersPolicy', {
          //responseHeadersPolicyName: ,
          comment: 'A default policy',
          securityHeadersBehavior: {
            //contentSecurityPolicy: { contentSecurityPolicy: 'default-src https:;', override: true },
            contentTypeOptions: { override: true },
            frameOptions: {
              frameOption: cloudfront.HeadersFrameOption.SAMEORIGIN,
              override: true,
            },
            referrerPolicy: {
              referrerPolicy: cloudfront.HeadersReferrerPolicy.STRICT_ORIGIN_WHEN_CROSS_ORIGIN,
              override: true,
            },
            strictTransportSecurity: {
              accessControlMaxAge: cdk.Duration.seconds(31536000), // 1 year
              preload: true,
              includeSubdomains: true,
              override: true,
            },
            xssProtection: {
              protection: true,
              modeBlock: true,
              override: true,
            },
          },
          customHeadersBehavior: {
            customHeaders: [
              {
                header: 'Cache-Control',
                value: 'no-cache, no-store, private',
                override: true,
              },
              {
                header: 'pragma',
                value: 'no-cache',
                override: true,
              },
            ],
          },
        }),
      },
      errorResponses: [
        {
          ttl: cdk.Duration.seconds(300),
          httpStatus: 403,
          responseHttpStatus: 200,
          responsePagePath: '/index.html',
        },
        {
          ttl: cdk.Duration.seconds(300),
          httpStatus: 404,
          responseHttpStatus: 200,
          responsePagePath: '/index.html',
        },
      ],
      priceClass: props.priceClass ?? cloudfront.PriceClass.PRICE_CLASS_ALL,
    });
  }
}