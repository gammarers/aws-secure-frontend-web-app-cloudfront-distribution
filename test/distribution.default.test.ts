import { Stack, App } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import * as acm from 'aws-cdk-lib/aws-certificatemanager';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { SecureFrontendWebAppCloudFrontDistribution } from '../src';

describe('SecureFrontendWebAppCloudFrontDistribution default testing', () => {
  const app = new App();
  const stack = new Stack(app, 'TestingStack');
  new SecureFrontendWebAppCloudFrontDistribution(stack, 'SecureFrontendWebAppCloudFrontDistribution', {
    certificate: new acm.Certificate(stack, 'Certificate', {
      domainName: 'example.com',
    }),
    domainName: 'example.com',
    originBucket: new s3.Bucket(stack, 'OriginBucket'),
  });

  const template = Template.fromStack(stack);

  it('Should match Cloudfront Distribution', async () => {

    template.hasResourceProperties('AWS::CloudFront::Distribution', Match.objectEquals({
      DistributionConfig: {
        Enabled: true,
        DefaultRootObject: 'index.html',
        HttpVersion: 'http2and3',
        IPV6Enabled: true,
        Aliases: [
          'example.com',
        ],
        CustomErrorResponses: Match.arrayWith([
          {
            ErrorCode: 403,
            ResponseCode: 200,
            ResponsePagePath: '/index.html',
            ErrorCachingMinTTL: 300,
          },
          {
            ErrorCode: 404,
            ResponseCode: 200,
            ResponsePagePath: '/index.html',
            ErrorCachingMinTTL: 300,
          },
        ]),
        DefaultCacheBehavior: {
          AllowedMethods: Match.arrayWith([
            'GET',
            'HEAD',
            'OPTIONS',
          ]),
          CachePolicyId: '658327ea-f89d-4fab-a63d-7e88639e58f6',
          CachedMethods: Match.arrayWith([
            'GET',
            'HEAD',
            'OPTIONS',
          ]),
          Compress: true,
          ViewerProtocolPolicy: 'https-only',
          ResponseHeadersPolicyId: {
            Ref: Match.stringLikeRegexp('ResponseHeadersPolicy.*'),
          },
          TargetOriginId: Match.stringLikeRegexp('TestingStackSecureFrontendWebAppCloudFrontDistributionOrigin.*'),
        },
        Origins: Match.arrayWith([
          {
            DomainName: {
              'Fn::GetAtt': [
                Match.stringLikeRegexp('OriginBucket.*'),
                'RegionalDomainName',
              ],
            },
            Id: Match.stringLikeRegexp('TestingStackSecureFrontendWebAppCloudFrontDistributionOrigin.*'),
            OriginAccessControlId: {
              'Fn::GetAtt': [
                Match.stringLikeRegexp('SecureFrontendWebAppCloudFrontDistributionOrigin1S3OriginAccessControl'),
                'Id',
              ],
            },
            S3OriginConfig: {
              OriginAccessIdentity: '',
            },
          },
        ]),
        PriceClass: 'PriceClass_All',
        ViewerCertificate: {
          AcmCertificateArn: {
            Ref: Match.stringLikeRegexp('Certificate.*'),
          },
          MinimumProtocolVersion: 'TLSv1.2_2021',
          SslSupportMethod: 'sni-only',
        },
      },
    }));
  });

  it('Should match Cloudfront ResponseHeadersPolicy', async () => {

    template.hasResourceProperties('AWS::CloudFront::ResponseHeadersPolicy', Match.objectEquals({
      ResponseHeadersPolicyConfig: {
        Name: Match.stringLikeRegexp('TestingStackResponseHeadersPolicy.*'),
        Comment: 'A default policy',
        CustomHeadersConfig: {
          Items: Match.arrayWith([
            {
              Override: true,
              Header: 'Cache-Control',
              Value: 'no-cache, no-store, private',
            },
            {
              Override: true,
              Header: 'pragma',
              Value: 'no-cache',
            },
          ]),
        },
        SecurityHeadersConfig: {
          ContentTypeOptions: {
            Override: true,
          },
          FrameOptions: {
            Override: true,
            FrameOption: 'SAMEORIGIN',
          },
          ReferrerPolicy: {
            Override: true,
            ReferrerPolicy: 'strict-origin-when-cross-origin',
          },
          StrictTransportSecurity: {
            Override: true,
            AccessControlMaxAgeSec: 31536000,
            IncludeSubdomains: true,
            Preload: true,
          },
          XSSProtection: {
            Override: true,
            ModeBlock: true,
            Protection: true,
          },
        },
      },
    }));
  });

  it('Should match snapshot.', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});
