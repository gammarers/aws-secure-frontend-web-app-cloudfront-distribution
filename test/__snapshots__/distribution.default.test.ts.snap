// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecureFrontendWebAppCloudFrontDistribution default testing Should match snapshot. 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Certificate4E7ABB08": {
      "Properties": {
        "DomainName": "example.com",
        "DomainValidationOptions": [
          {
            "DomainName": "example.com",
            "ValidationDomain": "example.com",
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestingStack/Certificate",
          },
        ],
        "ValidationMethod": "EMAIL",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "OriginBucketCA772B8F": {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "OriginBucketPolicyFD67BA59": {
      "Properties": {
        "Bucket": {
          "Ref": "OriginBucketCA772B8F",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Condition": {
                "StringEquals": {
                  "AWS:SourceArn": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":cloudfront::",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":distribution/",
                        {
                          "Ref": "SecureFrontendWebAppCloudFrontDistributionFBE118EA",
                        },
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "cloudfront.amazonaws.com",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "OriginBucketCA772B8F",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "ResponseHeadersPolicy13DBF9E0": {
      "Properties": {
        "ResponseHeadersPolicyConfig": {
          "Comment": "A default policy",
          "CustomHeadersConfig": {
            "Items": [
              {
                "Header": "Cache-Control",
                "Override": true,
                "Value": "no-cache, no-store, private",
              },
              {
                "Header": "pragma",
                "Override": true,
                "Value": "no-cache",
              },
            ],
          },
          "Name": "TestingStackResponseHeadersPolicyA6DAC75B",
          "SecurityHeadersConfig": {
            "ContentTypeOptions": {
              "Override": true,
            },
            "FrameOptions": {
              "FrameOption": "SAMEORIGIN",
              "Override": true,
            },
            "ReferrerPolicy": {
              "Override": true,
              "ReferrerPolicy": "strict-origin-when-cross-origin",
            },
            "StrictTransportSecurity": {
              "AccessControlMaxAgeSec": 31536000,
              "IncludeSubdomains": true,
              "Override": true,
              "Preload": true,
            },
            "XSSProtection": {
              "ModeBlock": true,
              "Override": true,
              "Protection": true,
            },
          },
        },
      },
      "Type": "AWS::CloudFront::ResponseHeadersPolicy",
    },
    "SecureFrontendWebAppCloudFrontDistributionFBE118EA": {
      "Properties": {
        "DistributionConfig": {
          "Aliases": [
            "example.com",
          ],
          "CustomErrorResponses": [
            {
              "ErrorCachingMinTTL": 300,
              "ErrorCode": 403,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
            {
              "ErrorCachingMinTTL": 300,
              "ErrorCode": 404,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
          ],
          "DefaultCacheBehavior": {
            "AllowedMethods": [
              "GET",
              "HEAD",
              "OPTIONS",
            ],
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "CachedMethods": [
              "GET",
              "HEAD",
              "OPTIONS",
            ],
            "Compress": true,
            "ResponseHeadersPolicyId": {
              "Ref": "ResponseHeadersPolicy13DBF9E0",
            },
            "TargetOriginId": "TestingStackSecureFrontendWebAppCloudFrontDistributionOrigin11EBA8679",
            "ViewerProtocolPolicy": "https-only",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2and3",
          "IPV6Enabled": true,
          "Origins": [
            {
              "DomainName": {
                "Fn::GetAtt": [
                  "OriginBucketCA772B8F",
                  "RegionalDomainName",
                ],
              },
              "Id": "TestingStackSecureFrontendWebAppCloudFrontDistributionOrigin11EBA8679",
              "OriginAccessControlId": {
                "Fn::GetAtt": [
                  "SecureFrontendWebAppCloudFrontDistributionOrigin1S3OriginAccessControl09AD5FFF",
                  "Id",
                ],
              },
              "S3OriginConfig": {
                "OriginAccessIdentity": "",
              },
            },
          ],
          "PriceClass": "PriceClass_All",
          "ViewerCertificate": {
            "AcmCertificateArn": {
              "Ref": "Certificate4E7ABB08",
            },
            "MinimumProtocolVersion": "TLSv1.2_2021",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "SecureFrontendWebAppCloudFrontDistributionOrigin1S3OriginAccessControl09AD5FFF": {
      "Properties": {
        "OriginAccessControlConfig": {
          "Name": "TestingStackSecureFrontendWeOrigin1S3OriginAccessControl1B89D02B",
          "OriginAccessControlOriginType": "s3",
          "SigningBehavior": "always",
          "SigningProtocol": "sigv4",
        },
      },
      "Type": "AWS::CloudFront::OriginAccessControl",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
